pkg/controllers/provisioning/scheduling/theoreticalnode.go

```diff
@@ -0,0 +1,137 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"errors"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/cloudprovider"
+	"github.com/aws/karpenter/pkg/utils/resources"
+)
+
+// TheoreticalNode is a set of constraints, compatible pods, and possible instance types that could fulfill
+// these constraints. This will be turned into one or more actual node instances within the cluster after bin packing.
+type TheoreticalNode struct {
+	Constraints           *v1alpha5.Constraints
+	PossibleInstanceTypes []cloudprovider.InstanceType
+	Pods                  []*v1.Pod
+}
+
+func NewTheoreticalNode(constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType) *TheoreticalNode {
+	return &TheoreticalNode{
+		PossibleInstanceTypes: instanceTypes,
+		Constraints:           constraints.DeepCopy(),
+	}
+}
+func (n TheoreticalNode) Compatible(pod *v1.Pod) error {
+	podRequirements := v1alpha5.NewPodRequirements(pod)
+	if err := n.Constraints.Requirements.Compatible(podRequirements); err != nil {
+		return err
+	}
+
+	// Ensure that at least one instance type of the instance types that we are already narrowed down to based on the
+	// existing pods can support the pod
+	matching := 0
+	for _, it := range n.PossibleInstanceTypes {
+		if n.isPodCompatible(pod, it) {
+			matching++
+		}
+	}
+
+	if matching == 0 {
+		return errors.New("no matching instance type found")
+	}
+	return nil
+}
+
+func (n *TheoreticalNode) Add(pod *v1.Pod) {
+	n.Pods = append(n.Pods, pod)
+	n.Constraints = n.Constraints.Tighten(pod)
+	var compatibleTypes []cloudprovider.InstanceType
+	for _, it := range n.PossibleInstanceTypes {
+		if n.isPodCompatible(pod, it) {
+			compatibleTypes = append(compatibleTypes, it)
+		}
+	}
+	n.PossibleInstanceTypes = compatibleTypes
+}
+
+// hasCompatibleNodeSelector returns true if a given set of node selectors match the instance type
+func (n TheoreticalNode) hasCompatibleNodeSelector(nodeSelector map[string]string, it cloudprovider.InstanceType) bool {
```

Comment:
Can we remove the binpacking validation now that we have this?

Reply:
Sounds good, I went ahead and moved all the logic to node and removed it from packable.